### ISSUE TITLE
Guidance on switching between GPU and CPU latency + memory usage

### ISSUE THREAD

**Issue by reiinakano** [PAIR-code/deeplearnjs#141]
Hi team, this is a bit of a long post, so please bear with me. I am currently working on a Fast Neural Style demo using Deeplearn.JS, and I'm happy to report that things have worked out well so far.

Here's [Scarlett Johansson](https://user-images.githubusercontent.com/18363734/30678063-4260c9b2-9ec2-11e7-940a-0065cd8cb82a.jpg) stylized with Udnie 

![Image of SJ](https://scontent.fmnl3-1.fna.fbcdn.net/v/t34.0-12/21905643_1628836347127520_304797976_n.png?_nc_eui2=v1%3AAeFAp-1TARhqrUbO1ohc05NyKGKuH1Wp1tgGYYGxtzWKUyQlKzlrKj7nqmTMxi2fFCSaHLvmPMB0m8UkVz583-WT71hRQ5QNvtCOmgERtWNMIUNO6vTjyGyUu5cjrr44GkY&oh=d62d59192ad21bc38c8461a364591b4c&oe=59C51C6B)

and the [Stata Center](https://scontent.fmnl3-1.fna.fbcdn.net/v/t34.0-0/p280x280/21903739_1628828280461660_1705293429_n.png?_nc_eui2=v1%3AAeGnDDkEJvsVfqfr6O5W7Od6eDvZjXUdTCaxX7cKN3UAcUXGxFk62SJolcB0yX6aqRSVFg9-v6z9f0EwBqMRN-jqlbwv9n2-07GJpz663b-Lf4Uu4KM2Kpa1WZDGjI8O4iA&oh=6dffe8af4dbab922da9b7270680a4e69&oe=59C62C94) stylized with Scream

![Stata](https://scontent.fmnl3-1.fna.fbcdn.net/v/t34.0-12/21905340_1628828520461636_1488600388_n.png?_nc_eui2=v1%3AAeFQpkpcgR9lY04VLXhP07cWKi4kWwzgxu4v69WQqgLCDEHO4IsDYUJPqahpe7JRRQE5o8J37lP1fnJo5_ZxV8V3IfksMXAHHzbfAne20IBkyybWn4IKR1KScnVcjtQ-VQM&oh=783db0b9f39eb1534c64919c05e3e561&oe=59C6328C)

Both happily produced by my browser :)

The problem is there's a pretty significant bottleneck in my implementation, which is the instance normalization in the model. Since I couldn't find a GPU version of `tf.nn.moments` (relevant issue at #138 ), I had to write my own in CPU (I'm not familiar enough to dive into WebGL stuff..).

Since the call to my `moments` function takes place 20+ times, with mathGPU functions in between, you can imagine the additional latency this causes.

Here's a cut down version of my `moments` function. I replaced lines of code with comments and highlighted only the parts I think are relevant to my issues. The call to this function happens multiple times inside my main mathGPU scope, interspersed between other mathGPU calls:
```ts
/**
   * Copies behavior of tf.nn.moments but purely for instance normalization.
   * Equivalent to tf.nn.moments(net, [0, 1], keep_dims=True) for a tensor 
   * of shape ()
   *
   * @param input Array3D shape [height, width, inDepth]
   * @return mean and variance per channel. Same shape as input.
   */
  private instanceMoments(input: Array3D): [Array3D, Array3D] {
    // Create explicit MathCPU for unimplemented GPU operations
    const mathCPU = new NDArrayMathCPU;

    // Switch dims for easier slicing. Operation is now on CPU
    const switched = mathCPU.switchDim(input, [2, 0, 1]);
    console.log('instanceMoments: switched dims');
    const switchedValues = switched.getValues();

    // Calculate Array keepDimMeans and keepDimVariances here.

    const meansArray = Array3D.new(input.shape, keepDimMeans);
    const variancesArray = Array3D.new(input.shape, keepDimVariances);

    return [meansArray, variancesArray];
  }
```

My problem aside from needing to switch to CPU, is it seems to be taking up a significant amount of memory (at peak it seems to take around 4GB of memory, forcing my crappy laptop to use SWAP). I compared it by using all MathCPU operations, and the memory kept steady for that one (although it took way longer). Am I doing something wrong? Is there a leak somewhere? Is this expected? 

Could you guide me on how to implement `tf.nn.moments` in GPU? It would simply need an implementation of `tf.reduce_mean` (with `keep_dims` option) in GPU.

If you would like to examine the full code of my `infer` function, it is here: https://github.com/reiinakano/deeplearnjs/blob/styletransfer-demo/demos/style-transfer/net.ts#L68

Thanks in advance!

**Comment by dsmilkov** [comment]
This is so cool to see and and great demo (saw on Twitter)! Sorry for the delay on our side. We'll prioritize and add the necessarily ops to make this run fast without gpu-cpu interlock. We can also help you connect your demo with the web-cam :) Stay tuned.

**Comment by reiinakano** [comment]
Thanks! It actually already has a webcam option. :)

No worries and take your time!

**Comment by nsthorat** [comment]
Ah I saw the webcam option, very very cool! Once we build the instance moments op we can do live style transfer with the webcam (instead of take a static picture => style transfer).

Also, apologies for retweeting some random person's link to your demo, I originally thought it must have been you. Once we do instance normalization and do webcam stuff we'll make sure to give you the official shout out! Really cool work!

**Comment by reiinakano** [comment]
No need to apologize, didn't bother me at all. Was just happy someone found it cool enough to tweet about it. :)

Hmm, idk about live style transfer, [the repository this implementation is based on](https://github.com/lengstrom/fast-style-transfer/) says it takes 100ms per frame on a Maxwell Titan X while several seconds on CPU. 

I'm assuming pure WebGL is closer to CPU Tensorflow? I just based that on the fact that this is what Tensorfire says on their website (also WebGL-based). Anyway, will look forward to how fast this can be with the necessary ops!

**Comment by nsthorat** [comment]
You won't be able to do it at 60fps, but you can just show style transfer frames from the webcam as fast as possible.

For those with beefy GPUs like a 1080 they'd see it at a much higher framerate!

**Comment by dsmilkov** [comment]
tf.nn.moments added in #277. Let us know if you run into any issues.

**Comment by reiinakano** [comment]
So I tried updating from 0.2.3 to 0.3.4 today without changing anything, and the demo works just fine on Firefox, but on my Chrome browser, the generated image is wrong.

Firefox (Correct):
![screenshot from 2017-11-03 22 11 03](https://user-images.githubusercontent.com/18363734/32377835-f0eabb44-c0e3-11e7-94e0-f969759edfef.png)

Chrome:
![screenshot from 2017-11-03 22 11 45](https://user-images.githubusercontent.com/18363734/32377865-017324ce-c0e4-11e7-8607-635f11d8cc97.png)

Any ideas? I can continue working on this with v0.3 using Firefox but I'd want to make it work on Chrome first before I deploy anything new.

**Comment by dsmilkov** [comment]
Interesting. We found that a useful way to debug is to print the first few values of the NDArray after each layer and see where they start to diverge (Firefox vs Chrome), given the same initial input/values. This can give us some details about which math operation is problematic.

In case I have some time to look at this, can you share the github branch that has your latest changes? Thanks!

**Comment by reiinakano** [comment]
Ahh, good idea, I'll do that.

Anyway, happy to report that I converted my code successfully using `math.moments` and it works well on Firefox. Here's the branch in case you want to check for yourself https://github.com/reiinakano/fast-style-transfer-deeplearnjs/tree/deeplearn-v0.3

**Comment by reiinakano** [comment]
Update: The difference has been pretty big with just the first `math.conv2d` call.

Firefox (presumably correct):
![screenshot from 2017-11-04 00 17 25](https://user-images.githubusercontent.com/18363734/32384464-efac6554-c0f5-11e7-93c0-9d3682e657b9.png)

Chrome:
![screenshot from 2017-11-04 00 17 20](https://user-images.githubusercontent.com/18363734/32384465-f13eec98-c0f5-11e7-818d-903c50e1ea41.png)

Sorry if the images are a bit small. Just check the values on the lower right side of both images.

**Comment by reiinakano** [comment]
Here's a better view of the problem:

Firefox:
![screenshot from 2017-11-04 00 41 55](https://user-images.githubusercontent.com/18363734/32385429-0de9cb62-c0f9-11e7-86e9-d055aba62d41.png)

Chrome:
![screenshot from 2017-11-04 00 41 58](https://user-images.githubusercontent.com/18363734/32385434-1209d0d4-c0f9-11e7-975b-58d7f2f8e4ac.png)

The first two operations are conv2d, followed by moments.
You can see from the console log that the values of conv2d and mean have already started to diverge, but are still slightly close. But you can also see that the variance calculations are wildly different. After this, there's no recovering for the rest of the calculations..

**Comment by dsmilkov** [comment]
Very interesting. I cloned your branch with `math.moments` and it works on my Chrome. I suspect the problem is related to a fallback of byte packing floats as ints, which should only happen for iOS/Safari, not Chrome.

From your Chrome, can you share the report you get on this page: http://webglreport.com/?v=2

Also if can you refresh your demo, each time using one of the 4 following query strings appended to the url:
```
?dljsflags=WEBGL_FLOAT_TEXTURE_ENABLED:false
?dljsflags=WEBGL_FLOAT_TEXTURE_ENABLED:true
?dljsflags=WEBGL_VERSION:1
?dljsflags=WEBGL_VERSION:2
```

and share the result with us (if it works or not). This query string overrides the environment regardless of what your browser supports.

Thank you!

**Comment by reiinakano** [comment]
?dljsflags=WEBGL_FLOAT_TEXTURE_ENABLED:false - doesn't work
?dljsflags=WEBGL_FLOAT_TEXTURE_ENABLED:true - works!
?dljsflags=WEBGL_VERSION:1 - doesn't work
?dljsflags=WEBGL_VERSION:2 - webgl_util.js:24 Uncaught Error: This browser does not support WebGL.(…) lol!

![screenshot from 2017-11-04 11 29 00](https://user-images.githubusercontent.com/18363734/32401884-63181af8-c153-11e7-9843-a083819eee9a.png)
![screenshot from 2017-11-04 11 29 03](https://user-images.githubusercontent.com/18363734/32401885-635690da-c153-11e7-824d-d5c3fddb73ab.png)

I guess I'm using some pretty dated hardware?..

**Comment by reiinakano** [comment]
Update: I managed to solve this by enabling this in `chrome://flags`

```
WebGL 2.0 Prototype Mac, Windows, Linux, Chrome OS, Android
Enabling this option allows web applications to access the experimental WebGL 2.0 feature. This option should only be used for application development, and should not be used when browsing arbitrary web sites. #enable-unsafe-es3-apisDisable
```

What worries me is this is still experimental in my browser? I saw some screenshots where this was simply "WebGL 2.0" and not "WebGL 2.0 Prototype".

Anyway, here is a screenshot of `chrome://gpu`. Has some problems that you might get info from.

![screenshot from 2017-11-04 16 45 22](https://user-images.githubusercontent.com/18363734/32403887-e7ffffde-c17f-11e7-8592-8eedeaf24f9a.png)

**Comment by dsmilkov** [comment]
Glad you got it working on your Chrome. WebGL 2.0 is definitely not experimental in Chrome ([see Can I Use](http://caniuse.com/#feat=webgl2), but might be on your machine, maybe due to outdated graphics drivers?

In any case, thanks for all the reports. We were able to identify the fundamental problem (overflow/underflow when packing floats as bytes), which means the style transfer won't work on iOS. The solution is not trivial but we will address it soon (within a month).

Thanks to your report, we also uncovered a bug in the environment feature testing code, which incorrectly resolved to packing floats as bytes whenever WebGL 2.0 was not available. We fixed it in #301 and released `0.3.6`. More specifically, your demo should now work in WebGL 1.0, i.e. with the flag `?dljsflags=WEBGL_VERSION:1`. If it doesn't, let us know!

Thanks!

**Comment by reiinakano** [comment]
Awesome, I updated it to 0.3.6 and I can now run it without any flags on WebGL 1.

The site is now up and running and is definitely faster with GPU moments. Guess I can consider this closed. Thanks for all the help!

**Comment by nsthorat** [comment]
I think Daniel has a couple other performance wins in store (like pooling the reduction ops so we get parallelization wins).

@reiinakano, what do you think of checking this demo into our `demos/` directory, then we can feature this on the homepage once Daniel finishes the performance optimization? You are welcome to put your name on the demo, of course, and when we tweet about it we'll credit you :)

**Comment by reiinakano** [comment]
@nsthorat Sure! I'll add it in a few days when I have time.
